File: core/src/net/natruid/jungle/utils/ai/actions/MoveTowardUnitAction.java
package net.natruid.jungle.utils.ai.actions;

import net.natruid.jungle.core.Sky;
import net.natruid.jungle.events.UnitMoveEvent;
import net.natruid.jungle.systems.TileSystem;
import net.natruid.jungle.utils.PathNode;
import net.natruid.jungle.utils.ai.BehaviorAction;
import net.natruid.jungle.utils.types.ExtractPathType;

import java.util.Deque;

public class MoveTowardUnitAction extends BehaviorAction {
    private final int preserveAp;

    private TileSystem tileSystem;
    private Deque<PathNode> path = null;

    public MoveTowardUnitAction(int preserveAp) {
        this.preserveAp = preserveAp;
    }

    public MoveTowardUnitAction() {
        this(0);
    }

    @Override
    public Float evaluate() {
        int target = getTargets().get(0);
        assert target >= 0;
        float maxMovement = unitManageSystem.getMovement(getSelf(), preserveAp);
        Deque<PathNode> path = pathfinderSystem.path(
            mUnit.get(getSelf()).tile,
            mUnit.get(target).tile,
            true,
            getSelf(),
            ExtractPathType.CLOSEST,
            maxMovement
        );
        if (path == null || path.isEmpty()) {
//            Sky.log.debug(getSelf() + " cannot find path");
            return null;
        }
        if (path.getLast().tile == mUnit.get(getSelf()).tile) {
//            Sky.log.debug(getSelf() + " is already at the destination");
            return null;
        }
        float newDist = tileSystem.getDistance(mUnit.get(target).tile, path.getLast().tile);
        this.path = path;
        return behaviorSystem.getScore(
            "move forward",
            unitManageSystem.getMovementCost(getSelf(), path.getLast().cost, true),
            -newDist
        );
    }

    @Override
    public boolean execute() {
        Sky.log.debug(getSelf() + " moving with path size: " + path.size());
        {
            UnitMoveEvent it = es.dispatch(UnitMoveEvent.class);
            it.unit = getSelf();
            assert path != null;
            it.path = path;
        }
        return true;
    }

    @Override
    public void reset() {
        path = null;
    }
}

File: core/src/net/natruid/jungle/utils/ai/actions/MoveTowardUnitAction.kt
package net.natruid.jungle.utils.ai.actions

import net.natruid.jungle.systems.TileSystem
import net.natruid.jungle.utils.ExtractPathType
import net.natruid.jungle.utils.Path
import net.natruid.jungle.utils.ai.BehaviorAction

class MoveTowardUnitAction(private val preserveAp: Int = 0) : BehaviorAction() {
    private lateinit var tileSystem: TileSystem

    private var path: Path? = null

    override fun evaluate(): Float? {
        val target = targets.first()
        assert(target >= 0)
        val maxMovement = unitManageSystem.getMovement(self, preserveAp)
        val path = pathfinderSystem.path(
            mUnit[self].tile,
            mUnit[target].tile,
            unit = self,
            type = ExtractPathType.CLOSEST,
            maxCost = maxMovement
        )
        if (path == null || path.isEmpty()) {
//            Logger.debug { "$self cannot find path" }
            return null
        }
        if (path.last.tile == mUnit[self].tile) {
//            Logger.debug { "$self is already at the destination" }
            return null
        }
        val newDist = tileSystem.getDistance(mUnit[target].tile, path.last.tile)
        this.path = path
        return behaviorSystem.getScore(
            "move forward",
            unitManageSystem.getMovementCost(self, path.last.cost, true),
            -newDist
        )
    }

    override fun execute(): Boolean {
//        Logger.debug { "$self moving with path size: ${path!!.size}" }
        unitManageSystem.moveUnit(self, path!!)
        mBehavior[self].fullMoveArea = null
        return true
    }

    override fun reset() {
        path = null
    }
}
